lua/lualine/components/diff.lua

```diff
@@ -0,0 +1,183 @@
+-- Copyright (c) 2020-2021 shadmansaleh
+-- MIT license, see LICENSE for more details.
+
+local async = require'lualine.async'
+local utils = require'lualine.utils.utils'
+local highlight = require"lualine.highlight"
+
+-- variable to store git diff stats
+local git_diff = nil
+
+-- process diff data and update git_diff{ added, removed, modified}
+local function process_diff(data)
+  -- Adapted from https://github.com/wbthomason/nvim-vcs.lua
+  local added, removed, modified = 0, 0, 0
+  for line in vim.gsplit(data, '\n') do
+    if string.find(line, [[^@@ ]]) then
+      local tokens = vim.fn.matchlist(line, [[^@@ -\v(\d+),?(\d*) \+(\d+),?(\d*)]])
+      local line_stats = {
+        mod_count = tokens[3] == '' and 1 or tonumber(tokens[3]),
+        new_count = tokens[5] == '' and 1 or tonumber(tokens[5])
+      }
+
+      if line_stats.mod_count == 0 and line_stats.new_count > 0 then
+        added = added + line_stats.new_count
+      elseif line_stats.mod_count > 0 and line_stats.new_count == 0 then
+        removed = removed + line_stats.mod_count
+      else
+        local min = math.min(line_stats.mod_count, line_stats.new_count)
+        modified = modified + min
+        added  = added + line_stats.new_count - min
+        removed = removed + line_stats.mod_count - min
+      end
+    end
+  end
+  git_diff = { added,  modified, removed }
+end
+
+-- variable to store git_diff getter async function
+local get_git_diff = nil
+-- flag to see if async job exited before updating git_diff
+local updated = false
+
+-- Updates the async function for current file
+local function update_git_diff_getter()
+  -- stop older function properly before overwritting it
+  if get_git_diff then get_git_diff:stop() end
+  -- Donn't show git diff when current buffer doesn't have a filename
+  if #vim.fn.expand('%') == 0 then get_git_diff = nil; git_diff=nil; return end
+  get_git_diff = async:new({
+    cmd = string.format([[git -C %s --no-pager diff --no-color --no-ext-diff -U0 -- %s]]
+    ,vim.fn.expand('%:h'), vim.fn.expand('%:t')),
+    on_stdout = function(_, data)
+      if data then
+        process_diff(data)
+        updated = true
+      end
+    end,
+    on_stderr = function (_, data)
+      if data then
+        git_diff = nil
+        updated = true
+      end
+    end,
+    on_exit = function()
+      if not updated then
+        -- updated not set means git exited without emmiting anything on stdout
+        -- or stderr means file is unchanged
+        git_diff = {0, 0, 0}
+        updated = true
+      end
+    end
+  })
+end
+
+-- Update git_diff veriable
+local function update_git_diff()
+  vim.schedule_wrap(function()
+    if get_git_diff then
+      updated = false
+      get_git_diff:start()
+    end
+  end)()
+end
+
+vim.api.nvim_exec([[
```

Comment:
shouldn't this be placed somewhere else? probably in the `signify` function. I don't like global calls.